### PR TITLE
change InputError to KeyError

### PR DESCRIPTION
**Pull request by Daafip**
Missed the Bart's last comment in #407 :||, this works better if we can actually see the sonar could test

### CODE REVIEW

**Comment by Daafip** [comment]
> I was hoping more for something like a webpage listing the basin_id's together with river/basin names, country, etc. I guess that doesn't exist?

Shall I also add a link to these maps with the basin_ids? https://github.com/Daafip/caravan-map

**Comment by BSchilperoort** [comment]
I was also working on fixing some things yesterday, see https://github.com/eWaterCycle/ewatercycle/commit/876efe8f693e8afe58e28e28b61b5a56f89480ae
Perhaps you can merge those changes. I was making the `get_basin_id` function a classmethod, as that's how it was described in the docstring.

> Shall I also add a link to these maps with the basin_ids? https://github.com/Daafip/caravan-map

Could be nice, yes. But perhaps better if you can host it on github pages. Perhaps moving it to the ewatercycle org might also be better. Also see the issue I opened https://github.com/Daafip/caravan-map/issues/1

**Comment by Daafip** [comment]
> But perhaps better if you can host it on github pages. Perhaps moving it to the ewatercycle org might also be better

https://www.ewatercycle.org/caravan-map/ Done! 

Do you have any suggestions for the index page?

**Comment by BSchilperoort** [comment]
> Do you have any suggestions for the index page?

I think if you copy part of the README there that would be good. Some explanation why the map exists (to explore and find basin IDs). And put in some relevant links to e.g. the ewatercycle documentation.

**Comment by Daafip** [comment]
> function a classmethod

How do I then mock  the function if its a class method?

**Comment by BSchilperoort** [comment]
> > function a classmethod
> 
> How do I then mock the function if its a class method?

I believe just with [mock.patch](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch).
https://stackoverflow.com/a/53485182

**Comment by Daafip** [comment]
> > > function a classmethod
> > 
> > 
> > How do I then mock the function if its a class method?
> 
> I believe just with [mock.patch](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch). https://stackoverflow.com/a/53485182

Ah, issue with my local unit testing and having the correct version

**Comment by Daafip** [comment]
Tried to fix the test coverage but don't think a higher coverage is easily met as most of uncovered is downloading files..

File: src/ewatercycle/_forcings/caravan.py [hunk 1]
```diff
@@ -300,8 +305,8 @@ def extract_basin_shapefile(
 
 def crop_ds(ds: xr.Dataset, start_time: str, end_time: str) -> xr.Dataset:
     """Crops dataset based on time."""
-    get_time(start_time), get_time(end_time)  # if utc, remove Z to parse to np.dt64
-    start, end = np.datetime64(start_time[:-1]), np.datetime64(end_time[:-1])
+    start = pd.Timestamp(start_time[:-1])
```
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: Perhaps it's more robust to use `get_time` from `ewatercycle.util` :thinking:

Also you don't need to remove the "Z" (Zulu = UTC+0). For pandas pd.Timestamp. However, then you get a timezone aware timestamp which can be more difficult.

File: tests/src/base/test_forcing.py [hunk 2]
```diff
@@ -331,6 +334,45 @@ def test_retrieve_caravan_forcing(tmp_path: Path, mock_retrieve: mock.MagicMock)
...
+
+def test_retrieve_caravan_forcing_no_basin_id(
+    tmp_path: Path, mock_retrieve: mock.MagicMock
+):
+    test_files_dir = Path(__file__).parent / "forcing_files"
+    tmp_camels_dir = tmp_path / "camels"
+    copytree(test_files_dir, tmp_camels_dir)
+
+    msg = (
+        "You have to specify a basin ID to be able to generate forcing from" " Caravan."
```
Comment (BSchilperoort) [inline tests/src/base/test_forcing.py]:
```suggestion
        "You have to specify a basin ID to be able to generate forcing from Caravan."
```
unnecessary string concatenation

**Review by BSchilperoort — APPROVED** [review]
Just two minor comments. Feel free to merge once these are resolved.

With the downloading writing sensible tests can be difficult. But I think this file is sufficiently covered now. The file has an 85% coverage, sonarcloud just doesn't like that some of the changed lines in this PR aren't covered (the measure is per PR, not per file...)

File: src/ewatercycle/_forcings/caravan.py (hunk 1 above)
Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: Was having issue with timezone aware mismatch...
If you do `pd.Timestamp(starttime)..`, the object you create is timezone aware whilst the forcing data is not...
Ill go with the get_time

Comment (BSchilperoort) [reply src/ewatercycle/_forcings/caravan.py]: I thought get_time would you give a timezone unaware datetime but not it's going wrong anyway? :/

Sorry for the confusion. I guess the original solution was quite OK, but it's probably good to parse out the "Z" differently, e.g., with `.replace("Z","")`.

Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: I went with `pd.Timestamp(get_time(start_time)).tz_convert(None)`
Makes it clear timezone awareness is dropped

**Pull request merged**